Initial implementation of BottomAppBar

amirh opened the pull request:
Per reviewers request, I'm sending a single big e2e CL and not sending the induvidual PRs for the scaffold, floating action button, and bottom app bar changes separately.

Currently there is no unit test for BottomAppBar, as I there is no easy way to make the bottom app bar and the floating action button overlap before #14368 lands. Will write tests once that CL lands.

TBD in following CLs:
- Unit test BottomAppBar.
- Add a flag to BottomAppBar to enable/disable the notch.
- Use default color for the BottomAppBar from the theme.
- Implicitly animate the BAB's color and elevation.

Hixie commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +189,52 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+
+  @override
+  void deactivate() {
+    Scaffold.setFloatingActionButtonNotchMakerFor(context, null);
```
does this work?

amirh replied:
yes

Hixie commented:
Seems pretty solid.

One problem with extendsWithPath rather than just passing the path is that it doesn't let you do things like have multiple parts to the path, e.g. you couldn't do something like:

```
-----.	 _____   .______
      \	 \___/  /
       \       /
       	\_____/
```

...but maybe that's ok.

amirh commented:
@Hixie if we need to support multi part paths in the future, could we add a method to ui.Path that concatenates path operations of 2 path objects? or is there some technical limitation or a principal reason to not do that?

amirh committed:
add a FAB NotchMaker to ScaffoldGeometry

amirh commented:
Actually - I see that we already have addPath that does that (right?), so switching to use it instead.

amirh committed:
add a notchMaker to FloatingActionButton

Hixie commented:
addPath won't do what you want (it ends the previous path).

I think extendWithPath is fine.

amirh committed:
Initial implementation of BottomAppBar

Mainly includes the notch making logic.
Not yet tested as currently there is no way to make the FAB and the BAB
overlap, once #14368 to lands we could add unit tests to the
BottomAppBar as well.

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+///   * [NotchMaker] a closure used for creating a notch in a shape.
+///   * [ScaffoldGeometry.floatingActionBarNotchMaker] the [NotchMaker] used to
+///   make a notch for the [FloatingActionButton]
+///   * [FloatingActionButton] which the [BottomAppBar] makes a notch for.
```
nit: for consistency with other dartdocs, only indent the bullets by one (asterisks aligned under the first "e") and wrap the text of these bullets so that they line up with the first character of their first line.

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+///   make a notch for the [FloatingActionButton]
+///   * [FloatingActionButton] which the [BottomAppBar] makes a notch for.
+class BottomAppBar extends StatefulWidget {
+  const BottomAppBar({
```
missing api docs; please check the output of `flutter analyze --flutter-repo --dartdocs` and use the prompts at https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-useless-documentation for ideas of what to write.

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+  /// The color to paint the material.
+  final Color color;
+
+  /// The z-coordinate at which to place this bototm app bar. This controls the
```
typo bottom

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+
+  @override
+  Widget build(BuildContext context) {
+      return new PhysicalShape(
```
overly indented.

(consider using `dartfmt` for formatting to avoid nits about formatting :-) )

amirh replied:
done.

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  @override
+  void didChangeDependencies() {
```
nit: this normally goes above the builder

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+    super.didChangeDependencies();
+    geometryListenable  = Scaffold.geometryOf(context);
+  }
+
```
nit: extraneous blank line

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    geometryListenable  = Scaffold.geometryOf(context);
```
nit: extraneous space before =

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+  Path getClip(Size size) {
+    final Rect babRect = Offset.zero & size;
+    if (geometry.value.floatingActionButtonArea == null ||
+      geometry.value.floatingActionButtonNotchMaker == null) {
```
nit: indent the expressions so they are vertically aligned

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+
+  @override
+  Path getClip(Size size) {
+    final Rect babRect = Offset.zero & size;
```
nit: avoid abbreviations

amirh replied:
went with appBar and button (for bottomAppBarRect and floatingActionButtonRect)

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+      return new Path()..addRect(babRect);
+    }
+
+    final Rect fab = geometry.value.floatingActionButtonArea
```
nit: avoid abbreviations

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,135 @@
+      .translate(0.0, geometry.value.bottomNavigationBarTop * -1);
+
+    final Rect intersect = babRect.intersect(fab);
+    if (intersect.width < 0 || intersect.height < 0) {
```
0.0

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +204,71 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+  Path _notchMaker(Rect host, Rect guest, Offset start, Offset end) {
+  if (end.dy != host.top)
+    throw new FlutterError(
+      'CircularTopNotchMaker must only be used for a notch in the top edge.\n'
```
Not sure what CircularTopNotchMaker is.

amirh replied:
Woops

Hixie commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +204,71 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+  }
+
+  Path _notchMaker(Rect host, Rect guest, Offset start, Offset end) {
+  if (end.dy != host.top)
```
indentation here seems off

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +204,71 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+  }
+
+  Path _notchMaker(Rect host, Rect guest, Offset start, Offset end) {
+  if (end.dy != host.top)
```
these checks should be in an assert block (or just make them asserts)

Hixie replied:
same applies to all the exception-throwing code below.

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -24,6 +24,21 @@ const double _kFloatingActionButtonMargin = 16.0; // TODO(hmuller): should be de
+/// The `host` is the bounding rectangle for the shape into which the notch will
+/// be applied.
+///
+///
```
extraneous blank lines

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -24,6 +24,21 @@ const double _kFloatingActionButtonMargin = 16.0; // TODO(hmuller): should be de
+///
+///
+/// The `host` is the bounding rectangle for the shape into which the notch will
+/// be applied.
```
you describe `host` but not `guest`?

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -24,6 +24,21 @@ const double _kFloatingActionButtonMargin = 16.0; // TODO(hmuller): should be de
+/// The `start` and `end` arguments are points on the outline of the host shape
+/// that will be connected by the returned path.
+///
+/// The returned path might pass inside the content bounds area.
```
this should describe the contract better. For example, can the path contain multiple subpaths? Can it end with a call to close()? How will it be used?

(If you require that it's one subpath and that it doesn't call close, consider using extendWithPath instead, and saying that subsequent subpaths are ignored.)

Hixie replied:
more contract questions: can the Path be mutated after it's returned? Can the path go outside the `guest` rectangle? outside the `host` rectangle? What order will `start` and `end` be in terms of the geometry? Will they always have the same `y` coordinate; will start always be to the left of end, etc.

amirh replied:
I added some docs.
The only guarantee for the return path is that it starts at `start` and ends at `end` and does not end with close(), anything else is allowed. I mentioned that, I emphasized that it can go inside the `guest` bounds, and that it can have multiple subpaths, as I think these are the things that might not be obvious. But I feel it is over verbose to enumerate a list of things the path is allowed to do...

Specifically the notchmaker used for the FAB can make further assumptions about the input. I added these assumptions to the documentation of `setFloatingActionButtonMakerFor`.

Regarding being able to mutate the path after it's returned, the answer is yes, but I do not see that we mention it in documentation for methods that returns a Path (e.g ShapeBorder.getOuterPath). I would expect the documentation to mention it only if the Path cannot be mutated.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -61,21 +77,37 @@ class ScaffoldGeometry {
   /// This is null when there is no floating action button showing.
   final Rect floatingActionButtonArea;
 
+  /// A [NotchMaker] for the floating action button.
```
maybe point to [NotchMaker] for the documentation of what the contract is

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -61,21 +77,37 @@ class ScaffoldGeometry {
+      );
+    }
 
     final Rect scaledFab = Rect.lerp(
```
nit: avoid abbrevations

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -107,11 +139,22 @@ class _ScaffoldGeometryNotifier extends ChangeNotifier implements ValueListenabl
+    notifyListeners();
+  }
+
+  void _updateFabNotchMaker(NotchMaker fabNotchMaker) {
```
nit: avoid abbreviations

amirh replied:
Done, though this one is really becoming too long `_updateFloatingActionButtonNotchMaker`, and is a private method so I would think it is reasonable to use FAB.

There's a tradeoff between including all the details everywhere and not being over verbose, and I'm not sure it is correct to always prefer to error on the more information side.

amirh replied:
Also we now have a `_setFloatingActionButtonNotchMakerAndNotify` method... it is only used in the same class, and called from 4 lines above it's declaration. I think using an abbreviation in this case is the right choice.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -230,6 +273,7 @@ class _ScaffoldLayout extends MultiChildLayoutDelegate {
         fabY = math.min(fabY, contentBottom - snackBarSize.height - fabSize.height - _kFloatingActionButtonMargin);
       if (bottomSheetSize.height > 0.0)
         fabY = math.min(fabY, contentBottom - bottomSheetSize.height - fabSize.height / 2.0);
+      fabY += 30.0;
```
?

amirh replied:
:open_mouth:
That what I did to test an app with some overlap

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -675,6 +719,20 @@ class Scaffold extends StatefulWidget {
+  ///
+  /// Returns true if the [ScaffoldGeometry.floatingActionButtonNotchMaker] was
+  /// set.
+  /// Returns false if there as no [Scaffold] ancestor.
```
this should talk about the contract, e.g. that you must call this with null if you are deactivated.

...having said that, I just thought of a problem with the setting to null.

Imagine you replace the scaffold's fab slot is a column with two children, and the fab is in the second child, and then you update the scaffold so that it's fab slot is a column with two children, but the fab is now in the first child.

The new fab will be built and will call `setFloatingActionButtonNotchMakerFor(this...)`, then the old fab will be deactivated and will call `setFloatingActionButtonNotchMakerFor(null)`.

amirh replied:
Replaced this with a closeable pattern. And extended the docs.

amirh committed:
use a closeable for clearing the FAB notchmaker

amirh commented:
@Hixie PTAL

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,139 @@
+///  * [NotchMaker] a closure used for creating a notch in a shape.
+///  * [ScaffoldGeometry.floatingActionBarNotchMaker] the [NotchMaker] used to
+///  make a notch for the [FloatingActionButton]
+///  * [FloatingActionButton] which the [BottomAppBar] makes a notch for.
```
AppBar, for an equivalent widget but shown at the top of the application.

amirh replied:
Added: `[AppBar] for a toolbar that is shown at the top of the screen.`
As these widget are not very equivalent (BottomAppBar is taking care of the notch, while AppBar is taking care of layout).

Hixie commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -117,13 +126,24 @@ class FloatingActionButton extends StatefulWidget {
 class _FloatingActionButtonState extends State<FloatingActionButton> {
   bool _highlight = false;
 
+  VoidCallback clearNotchMaker;
```
either this should be private, or highlight should be public, but we shouldn't have both private and public members on this State.

Whichever you decide to do, do the same for _BottomAppBarState

Hixie replied:
(i think convention is to make these private, but it doesn't much matter since they are private classes)

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +206,81 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+    assert(() {
+      if (end.dy != host.top)
+        throw new FlutterError(
+          'the floating action button\'s notch maker must only be used for a notch in the top edge.\n'
```
should start with an uppercase letter (also applies to other error text below)

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +206,81 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+      if (end.dy != host.top)
+        throw new FlutterError(
+          'the floating action button\'s notch maker must only be used for a notch in the top edge.\n'
+          'end point: $end is not in the top edge of $host'
```
i think if i saw this error message i would not be sure what "end point" means here.

Hixie replied:
(similar comment for the next error)

amirh replied:
Change to "The notch's end point". Not sure if it's much better...
Though if you've seen this error it probably means that you are directly using a notch maker, so you should probably have some context about the parameters.

Hixie commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +206,81 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+    final Rect intersection = host.intersect(guest);
+    assert(() {
+      if (intersection.width < 0 || intersection.height < 0)
+        throw new FlutterError('notch host must intersect with it\'s guest');
```
its

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -24,6 +24,24 @@ const double _kFloatingActionButtonMargin = 16.0; // TODO(hmuller): should be de
+/// The `start` and `end` arguments are points on the outline of the host shape
+/// that will be connected by the returned path.
+///
+/// The returned path might pass anyware including inside the guest bounds area,
```
might -> may
anyware -> anywhere
add a comma after "anywhere" unless you actually mean "may pass only inside the guest bounds area"

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -24,6 +24,24 @@ const double _kFloatingActionButtonMargin = 16.0; // TODO(hmuller): should be de
+///
+/// The returned path might pass anyware including inside the guest bounds area,
+/// and may contain multiple subpaths. The returned path ends at `end` and does
+/// not end with a [Path.close]. The returned Path is built under the
```
being allowed to contain multiple subpaths seems contradictory with not being allowed to end with Path.close?

amirh replied:
Why is that? you can have Path.close, just don't end with it...

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -24,6 +24,24 @@ const double _kFloatingActionButtonMargin = 16.0; // TODO(hmuller): should be de
+///
+/// The returned path might pass anyware including inside the guest bounds area,
+/// and may contain multiple subpaths. The returned path ends at `end` and does
+/// not end with a [Path.close]. The returned Path is built under the
```
[Path]

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -61,32 +80,68 @@ class ScaffoldGeometry {
+class _Closeable {
+  VoidCallback closeCallback;
+
+  _Closeable(this.closeCallback) : assert(closeCallback != null);
```
constructors first

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -100,18 +155,36 @@ class _ScaffoldGeometryNotifier extends ChangeNotifier implements ValueListenabl
+    notchMakerCloseable?.close();
+    _setFloatingActionButtonNotchMakerAndNotify(fabNotchMaker);
+    notchMakerCloseable = new _Closeable(() { _setFloatingActionButtonNotchMakerAndNotify(null); });
+    return notchMakerCloseable.close;
```
This has the same bug as the previous code before the closeable, as far as i can tell.

Hixie replied:
oh, never mind, i see.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -675,6 +748,29 @@ class Scaffold extends StatefulWidget {
+  /// [Scaffold] ancestor of the given context if one exist.
+  ///
+  /// It is guaranteed that `notchMaker` will only be used for making notches
+  /// in the top edge of the [BottomAppBar], the start and end offsets given to
```
rather than linking to [BottomAppBar], I think it would be strictly correct to link to [bottomNavigationBar].

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -675,6 +748,29 @@ class Scaffold extends StatefulWidget {
+  ///
+  /// Callers must invoke the callback when the notch is no longer required.
+  /// This method is typically called from [State.didChangeDependencies] and the
+  /// callback should then be invoked from [State.deactivate].
```
This should probably mention that only the last call will be remembered, so if there's two FloatingActionButtons, only one will be represented in the ScaffoldGeometry.

amirh replied:
Added. Though it feels a bit over-verbose to mention that a setX method overrides the previous value of X...

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -675,6 +748,29 @@ class Scaffold extends StatefulWidget {
     return scaffoldScope.geometryNotifier;
```
Add BottomAppBar to the list of "See also"s for Scaffold around line 500.

amirh replied:
done

amirh commented:
Addressed all comments.
cc @HansMuller as Ian is out this week

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -675,6 +750,32 @@ class Scaffold extends StatefulWidget {
   }
 
+  /// Sets the [ScaffoldGeometry.floatingActionButtonNotchMaker] for the closest
+  /// [Scaffold] ancestor of the given context if one exist.
```
context if one exist => context, if one exists

amirh replied:
done

HansMuller reviewed (commented):
This looks great! Just some minor tweaks and suggestions.

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+// Examples can assume:
+// Widget bottomAppBarContents;
+
+/// A container that shows in the [Scaffold.bottomNavigationBar] slot, and can
```
done

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+// Widget bottomAppBarContents;
+
+/// A container that shows in the [Scaffold.bottomNavigationBar] slot, and can
+/// have a notch in which a [FloatingActionButton] is showing.
```
done

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -24,6 +24,24 @@ const double _kFloatingActionButtonMargin = 16.0; // TODO(hmuller): should be de
+/// not end with a [Path.close]. The returned [Path] is built under the
+/// assumption it will be added to an existing path that is at the `start`
+/// coordinates using [Path.addPath].
+typedef Path NotchMaker(Rect host, Rect guest, Offset start, Offset end);
```
I wasn't very happy with `NotchPathBuilder` as I don't feel it's a builder (doesn't have a build method).

After f2f we concluded to  replace `NotchMaker` with `ComputeNotch` and `ScaffoldGeometry.FloatingActionButtonNotchMaker` with `ScaffoldGeometry.FloatingActionButtonNotch`

amirh committed:
rename NotchMaker to ComputeNotch

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  /// The color to paint the material.
```
done.

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+  const BottomAppBar({
+    Key key,
+    this.child,
+    // TODO(amirh): use a default color from the theme.
```
done

amirh commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +206,81 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+    // center of the notch and the intersection of the notch's circle and the host's
+    // top edge.
+    //
+    // The hypotenuse of this triangle equals to the notch's radius, and one side
```
done

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -24,6 +24,24 @@ const double _kFloatingActionButtonMargin = 16.0; // TODO(hmuller): should be de
+/// be applied. The `guest` is the bounding rectangle of the shape for which we
+/// are creating a notch in the host.
+///
+/// The `start` and `end` arguments are points on the outline of the host shape
```
For the generic notch we don't need to require that they are necessarily within the host bounds. 
For the fab notch we have some extra guarantees that are described in the doc for setFloatingActionButtonNotch.
It says `the start and end offsets ... will always be on the top edge of [bottomNavigationBar]` (that should imply they are between host.topLeft and host.topRight, happy to also explicitly say that there if you think it clarifies).

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    final Rect button = geometry.value.floatingActionButtonArea
+      .translate(0.0, geometry.value.bottomNavigationBarTop * -1);
```
done, hope I caught everything (I did /\d[^.] on bottom_app_bar.dart and floating_action_button.dart)

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -24,6 +24,24 @@ const double _kFloatingActionButtonMargin = 16.0; // TODO(hmuller): should be de
 
+/// Returns a path for a notch in the outline of a shape.
+///
+/// The path makes a notch in the host shape that can contain the guest shape.
```
They are explained 2 paragraphs below, did you mean to change the paragraph order?

amirh committed:
review comments followup

amirh commented:
Thanks for the review!
I think I addressed all comments, PTAL.

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+  /// The [color] and [elevation] arguments must not be null.
+  const BottomAppBar({
+    Key key,
+    this.child,
```
done

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+  @override
+  Widget build(BuildContext context) {
+    return new PhysicalShape(
+      child: new Material(
```
done

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -46,6 +64,7 @@ class ScaffoldGeometry {
   const ScaffoldGeometry({
     this.bottomNavigationBarTop,
     this.floatingActionButtonArea,
+    this.floatingActionButtonNotchMaker,
```
Used floatingActionButtonNotch

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+    final Rect button = geometry.value.floatingActionButtonArea
+      .translate(0.0, geometry.value.bottomNavigationBarTop * -1);
+
+    final Rect intersect = appBar.intersect(button);
```
thanks

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+      return new Path()..addRect(appBar);
+    }
+
+    final Rect button = geometry.value.floatingActionButtonArea
```
done

amirh commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -35,6 +39,10 @@ class _DefaultHeroTag {
 /// If the [onPressed] callback is null, then the button will be disabled and
 /// will not react to touch.
 ///
+/// If the floating action bar is a descendant of a [Scaffold] that also has a
```
done

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+///
+/// See also:
+///
+///  * [NotchMaker] a closure used for creating a notch in a shape.
```
done

amirh commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -2,12 +2,15 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:math';
```
done

amirh commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -35,6 +39,10 @@ class _DefaultHeroTag {
 ///
+/// If the floating action bar is a descendant of a [Scaffold] that also has a
+/// [BottomAppBar], the [BottomAppBar] will show a notch to accomodate the
+/// [FloatingActionButton] when it overlaps the [BottomAppBar].
```
done. I did not mention that it can be changed with `Scaffold.setFloatingActionButtonNotch` as I think the users of `FloatingActionButton` shouldn't care about how it tells the scaffold about the notch...

amirh commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -117,13 +126,24 @@ class FloatingActionButton extends StatefulWidget {
 
+  /// The margin to keep around the floating action button when creating a
+  /// notch for it.
+  ///
```
added a paragraph, see if you like it better

amirh commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -117,13 +126,24 @@ class FloatingActionButton extends StatefulWidget {
+  /// See also:
+  ///
+  ///   * [BottomAppBar], a material design elements that shows a notch for the
+  ///   floating action button.
```
done

amirh commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +206,81 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+    assert(() {
+      if (end.dy != host.top)
+        throw new FlutterError(
+          'The floating action button\'s notch maker must only be used for a notch in the top edge.\n'
```
done

amirh commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +206,81 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+
+    final Rect intersection = host.intersect(guest);
+    assert(() {
+      if (intersection.width < 0 || intersection.height < 0)
```
done

amirh commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -186,4 +206,81 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+      return true;
+    }());
+
+    // The FAB's shape is a circle bound by the guest rectangle.
```
done

amirh committed:
review comments followup

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+    @required this.color,
+    this.elevation: 8.0,
+  }) : assert(color != null),
+       assert(elevation != null),
```
done

amirh commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -0,0 +1,140 @@
+///
+///  * [NotchMaker] a closure used for creating a notch in a shape.
+///  * [ScaffoldGeometry.floatingActionBarNotchMaker] the [NotchMaker] used to
+///  make a notch for the [FloatingActionButton]
```
done

amirh committed:
more nites

amirh committed:
follow up

HansMuller commented:
LGTM
